packages/client/src/util/fetch.ts

```diff
@@ -11,3 +26,82 @@ export function getFetchImplementation(userFetch?: FetchImpl) {
   }
   return fetchImpl;
 }
+
+export class ApiRequestPool {
+  #fetch?: FetchImpl;
+  #queue: Array<(...params: any[]) => any>;
+  #concurrency: number;
+
+  running: number;
+  started: number;
+
+  constructor(concurrency = 10) {
+    this.#queue = [];
+    this.#concurrency = concurrency;
+
+    this.running = 0;
+    this.started = 0;
+  }
+
+  setFetch(fetch: FetchImpl) {
+    this.#fetch = fetch;
+  }
+
+  getFetch(): FetchImpl {
+    if (!this.#fetch) {
+      throw new Error('Fetch not set');
+    }
+
+    return this.#fetch;
+  }
+
+  request(url: string, options?: RequestInit, context = { start: new Date(), stalled: false }): Promise<Response> {
+    const fetch = this.getFetch();
+
+    return this.#enqueue(async () => {
+      const response = await fetch(url, options);
+
+      if (response.status === 429) {
+        const rateLimitReset = parseNumber(response.headers?.get('x-ratelimit-reset')) ?? 1;
+
+        await timeout(rateLimitReset * 1000);
+        return await this.request(url, options, { ...context, stalled: true });
+      }
+
+      if (context.stalled) {
+        const stalledTime = new Date().getTime() - context.start.getTime();
+        console.warn(`A request to Xata hit your workspace limits, was retried and stalled for ${stalledTime}ms`);
+      }
+
+      return response;
+    });
+  }
+
+  #enqueue<Result>(task: () => Promise<Result> | Result): Promise<Result> {
+    const promise = new Promise<Result>((resolve) => this.#queue.push(resolve))
+      .finally(() => {
+        this.started--;
+        this.running++;
+      })
+      .then(() => task())
+      .finally(() => {
+        this.running--;
+
+        const next = this.#queue.shift();
+        if (next !== undefined) {
+          this.started++;
+          next();
+        }
+      });
+
+    if (this.#queue.length > 0 && this.running + this.started < this.#concurrency) {
+      const next = this.#queue.shift();
+      if (next !== undefined) {
+        this.started++;
+        next();
+      }
+    }
```

Comment:
I think we can remove the `this.#queue.length > 0` check since you are already checkin if `next !== undefined` later.

Nitpick: lines 98-102 are repeated above. We could extract a function `runNext()` and call it twice. WDYT?